File: 프로그래머스/lv3/72414. 광고 삽입/광고 삽입.py
import time
from collections import Counter
def int_to_strtime(time):
    if time < 10:
        time = '0'+str(time)
    else: time = str(time)
    return time

def strtime(sec):
    h = sec // 3600
    m = sec % 3600
    s = m % 60
    m = m // 60
    return ':'.join([int_to_strtime(h), int_to_strtime(m), int_to_strtime(s)])

def convert(time):
    h, m, s = map(int, time.split(":"))
    s += m * 60
    s += h * 60 * 60
    return s

def solution(play_time, adv_time, logs):
    play_time = convert(play_time)
    adv_time = convert(adv_time)
    dp = [0 for _ in range(play_time + 1)]
    
    for l in logs:
        start, end = l.split("-")
        dp[convert(start)] += 1
        dp[convert(end)] -= 1
    
    for i in range(len(dp)):
        dp[i] = dp[i]+dp[i-1]
    
    viewer = 0
    maxx = dp[adv_time] - dp[0]
    start_time = 0
    for i in range(play_time-adv_time):
        viewer += dp[i+adv_time] - dp[i]
        if maxx < viewer:
            start_time = i + 1
            maxx = viewer
    
    answer = strtime(start_time)
    return answer

File: 프로그래머스/lv3/72414. 광고 삽입/광고 삽입.kt
import kotlin.math.*

fun formating(time:Int):String{
    if (time < 10){
        return "0"+time.toString()
    }else return time.toString()
}

fun convert(time:String):Int{
    var t = time.split(":")
    var s = t[0].toInt() * 3600
    s += t[1].toInt() * 60
    s += t[2].toInt()
    return s
}

fun strTime(time:Int):String{
    var h = time / 3600
    var m = time % 3600 / 60
    var s = time % 3600 % 60
    
    var strtime = arrayOf(formating(h), formating(m), formating(s))
    return strtime.joinToString(":")
}

class Solution {
    fun solution(play_time: String, adv_time: String, logs: Array<String>): String {
        var play = convert(play_time)
        var adv = convert(adv_time)
        
        var dp = Array(play+1,{0L})
        
        for (log in logs){
            var l = log.split("-")
            dp[convert(l[0])] += 1L
            dp[convert(l[1])] -= 1L
        }
        
        for (i in 1 until dp.size){
            dp[i] += dp[i-1]
        }
        
        for (i in 1 until dp.size){
            dp[i] += dp[i-1]
        }
        
        
        var viewer = dp[adv]-dp[0]
        var min_time = 0
        for (i in 0 until play - adv){
            if (viewer< dp[i+adv]-dp[i]){
                min_time = i+1
                viewer = dp[i+adv]-dp[i]
            }
        }
        
        return strTime(min_time)
    }
}